kaldo/controllers/isotopic.py

```diff
@@ -0,0 +1,194 @@
+        overlap=np.abs(overlap)**2
+        # print(eigvectors_.shape,overlap.shape,g_factor.shape)
+        g_per_mode=contract('ni,i->n',overlap,g_factor)
+        w2delta=omegas[condition]**2*curve(delta_omega[condition],2*np.pi*sigma)
+        bw=w2delta*g_per_mode/n_k_points
+        isotopic_bw[index_k,mu]=(np.pi/2)*np.sum(bw)
+
+    return isotopic_bw
+
+
+def calculate_base_sigma(velocity, cellinv, k_size):
+    #sigma: array (nk,nmodes)
+    #local adaptive broadening from Shengbte
+    # we want the last index of velocity (the coordinate index to dot from the right to rlattice vec
+    delta_k =np.dot( cellinv, 1/ k_size)
+    base_sigma = (contract('knx,x->kn',velocity,delta_k))**2
+    base_sigma = np.sqrt(base_sigma/6 )
+    return base_sigma
+
+def refine_sigma(base_sigma):
+    #sigma: array (nk,nmodes)
+    #local adaptive broadening similar to Shengbte
+    #avoid sigma too extreme ( e.g. zero)
+    sigma=base_sigma.copy()
+    sigma[base_sigma<=0]=np.min(sigma[base_sigma>0])
+    logsigma=np.log(sigma)
+    per25=np.percentile(logsigma, 25)
+    per50 = np.percentile(logsigma, 50)
+    per75 = np.percentile(logsigma, 75)
+    # delta=per75-per25
+    # lbound=np.exp(per25-1.5*delta)
+    lbound=np.exp(per75)
+    sigma=np.where(sigma>lbound,sigma,lbound)
+    logging.info('per25,per50,per75,mean sigma={} {} {} {}'.format(np.exp(per25),np.exp(per50),\
+                                                                np.exp(per75),np.mean(sigma)) )
+
+    return sigma
+
+def compute_gfactor(list_of_atomic_numbers):
+    g_factor=np.zeros(len(list_of_atomic_numbers))
+    minimal_list=np.unique(list_of_atomic_numbers)
+    try:
+        isotopes = download_isotope_data()
+        logging.info('downloading isotopic data from NIST database, using ase.data.isotopes.')
+        for element in minimal_list:
+            masses = np.array([isotopes[element][iso]['mass'] for iso in isotopes[element].keys()])
+            conc = np.array([isotopes[element][iso]['composition'] for iso in isotopes[element].keys()])
+            m_avg = np.sum(masses * conc)
+            rel_masses = masses / m_avg
+            g_ = np.sum(conc * (1 - rel_masses) ** 2)
+            g_factor[list_of_atomic_numbers == element] = g_
+    except:
+        g_factor_dict=legacy_database()
+        logging.info('online isotopic data not available, using legacy data.')
+        for element in minimal_list:
+            g_factor[list_of_atomic_numbers == element] = g_factor_dict[element]
+
+
+    return g_factor
+
+
+def legacy_database():
+    ## Legacy gfactor database. The isotopic database was downloaded with ase.data.isotopes on 20/03/2024.
+    # unstable elements have None as gfactor. Mostly elements with Z>92
+    database={1: 0.00011460742533846188, 2: 8.141017510851699e-08, 3: 0.001458820047621947, 4: 0.0, 5: 0.0013539153245004177, 6: 7.387230454537788e-05, 7: 1.8376657914224594e-05, 8: 3.3588025065260784e-05, 9: 0.0, 10: 0.0008278312160539562,\
```

Comment:
MInor, can you move this `database`  to a different file like a `isotopes.csv`.

Reply:
Good idea, I just moved the database in a json file kaldo/controllers/legacy_dataset.json